code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,364 @@
+        if (services.isEmpty()) {
+            return;
+        }
+
+        lockManager.lock(createLock(services), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                final List<DeploymentUnit> units = collectDeploymentUnits(services);
+                final int requestedScale = DataAccessor.fieldInteger(services.get(0),
+                        ServiceDiscoveryConstants.FIELD_SCALE);
+                // don't process if there is no need to reconcile
+                boolean needToReconcile = needToReconcile(services, units, requestedScale);
+
+                if (!needToReconcile) {
+                    return;
+                }
+                activateServices(service, services);
+                activateDeploymentUnits(services, units, requestedScale);
+            }
+        });
+    }
+
+    private boolean needToReconcile(final List<Service> services, final List<DeploymentUnit> units, final int requestedScale) {
+        for (Service service : services) {
+            if (service.getState().equals(CommonStatesConstants.INACTIVE)) {
+                return true;
+            }
+        }
+
+        boolean needToReconcile = false;
+        if (units.size() == requestedScale) {
+            for (DeploymentUnit unit : units) {
+                if (unit.isError()) {
+                    needToReconcile = true;
+                    break;
+                }
+                if (unit.getInstances().size() != services.size()) {
+                    needToReconcile = true;
+                    break;
+                }
+                if (!unit.isStarted()) {
+                    needToReconcile = true;
+                    break;
+                }
+                if (unit.needsCleanup()) {
+                    needToReconcile = true;
+                    break;
+                }
+            }
+        } else {
+            needToReconcile = true;
+        }
+        return needToReconcile;
+    }
+
+    private void activateServices(final Service initialService, final List<Service> services) {
+        /*
+         * Trigger activate for all the services
+         */
+        try {
+            for (Service service : services) {
+                if (service.getState().equalsIgnoreCase(CommonStatesConstants.INACTIVE)) {
+                    objectProcessManager.scheduleStandardProcess(StandardProcess.ACTIVATE, service, null);
+                } else if (service.getState().equalsIgnoreCase(CommonStatesConstants.ACTIVE)) {
+                    objectProcessManager.scheduleStandardProcess(StandardProcess.UPDATE, service, null);
+                }
+            }
+        } catch (IdempotentRetryException ex) {
+            // if not caught, the process will keep on spinning forever
+            // figure out better solution
+        }
+
+    }
+
+    protected LockDefinition createLock(List<Service> services) {
+        return new ServicesSidekickLock(Sets.newHashSet(services));
```

Comment:
Just change the constructor to `Collection<Service>` and then you don't have to convert